src/Parser.php

```diff
@@ -93,38 +158,44 @@ public function getIterator()
         ${','} = self::COMMA;
         ${':'} = self::COLON;
 
-        $iteratorLevel = count($this->jsonPointerPath);
         $iteratorStruct = null;
         $currentPath = [];
-        $pathFound = false;
+        $pathsFound = [];
         $currentLevel = -1;
         $stack = [$currentLevel => null];
         $jsonBuffer = '';
         $key = null;
-        $previousToken = null;
         $objectKeyExpected = false;
         $inObject = true; // hack to make "!$inObject" in first iteration work. Better code structure?
         $expectedType = self::OBJECT_START | self::ARRAY_START;
         $subtreeEnded = false;
         $token = null;
+        $currentPathChanged = true;
+        $jsonPointerPath = [];
+        $iteratorLevel = 0;
 
         // local variables for faster name lookups
         $lexer = $this->lexer;
-        $jsonPointerPath = $this->jsonPointerPath;
-        
+
         foreach ($lexer as $token) {
+            if ($currentPathChanged) {
+                $currentPathChanged = false;
+                $jsonPointerPath = $this->getMatchingJsonPointerPath($currentPath);
```

Comment:
I strive to call as few functions as possible in this loop. Also, the function seems overly complex to do the job. To compare current path against multiple JSON pointers I was thinking something along the lines of `isset($jsonPointers[$currentPath]) || isset($jsonPointers[preg_replace('~/[0-9]+[/$]~', '-', $currentPath)])` (treat is as pseudocode to put an idea across).

Reply:
The function is that complex because we also need the closest match in case there is no full match yet, otherwise it would discard a JSON pointer path too soon. I've tried a simpler approach using RegExp, but so far I could not find any other functional solution.